### PR TITLE
composite primary key support for SF, PG and BQ

### PR DESCRIPTION
**Pull request by heavycrystal**
Fixes issues: #491 #492 #493 #494 #495 
Supersedes: #129 and #164

### CODE REVIEW

File: flow/activities/flowable.go [hunk 1]
```diff
@@ -358,10 +358,12 @@ func (a *FlowableActivity) StartNormalize(
 		return nil, fmt.Errorf("failed to normalized records: %w", err)
 	}
 
-	err = a.CatalogMirrorMonitor.UpdateEndTimeForCDCBatch(ctx, input.FlowConnectionConfigs.FlowJobName,
-		res.EndBatchID)
-	if err != nil {
-		return nil, err
+	if res.Done {
```
Comment (iskakaushik) [inline flow/activities/flowable.go:361]: can you add a comment

File: flow/cmd/handler.go [hunk 2]
```diff
@@ -100,7 +100,7 @@ func (h *FlowRequestHandler) CreateCDCFlow(
 
 	maxBatchSize := int(cfg.MaxBatchSize)
 	if maxBatchSize == 0 {
-		maxBatchSize = 100000
+		maxBatchSize = 200000
```
Comment (iskakaushik) [inline flow/cmd/handler.go:103]: revert it back

File: flow/connectors/bigquery/bigquery.go [hunk 3]
```diff
@@ -1288,14 +1288,12 @@ func (m *MergeStmtGenerator) generateDeDupedCTE() string {
 			) _peerdb_ranked
 			WHERE _peerdb_rank = 1
 	) SELECT * FROM _peerdb_de_duplicated_data_res`
-	pkey := m.NormalizedTableSchema.PrimaryKeyColumn
-	return fmt.Sprintf(cte, pkey)
+	pkeyColsStr := fmt.Sprintf("(CONCAT(%s))", strings.Join(m.NormalizedTableSchema.PrimaryKeyColumns, ", ' ', "))
```
Comment (iskakaushik) [inline flow/connectors/bigquery/bigquery.go:1291]: make it something like `_peerdb_concat_`

File: flow/connectors/postgres/cdc.go [hunk 4]
```diff
@@ -617,3 +605,18 @@ func (p *PostgresCDCSource) processRelationMessage(
...
+
+	for _, pkeyCol := range req.TableNameSchemaMapping[tableName].PrimaryKeyColumns {
+		pkeyColVal, err := rec.GetItems().GetValueByColName(pkeyCol)
+		if err != nil {
+			return "", fmt.Errorf("error getting pkey column value: %w", err)
+		}
+		pkeyColsMerged = append(pkeyColsMerged, fmt.Sprintf("%v", pkeyColVal.Value))
+	}
+
+	return strings.Join(pkeyColsMerged, " "), nil
```
Comment (iskakaushik) [inline flow/connectors/postgres/cdc.go:621]: hash it instead.

File: flow/connectors/postgres/cdc.go [hunk 5]
```diff
@@ -240,53 +241,42 @@ func (p *PostgresCDCSource) consumeStream(
 					// tableName here is destination tableName.
 					// should be ideally sourceTableName as we are in PullRecords.
 					// will change in future
-					isFullReplica := req.TableNameSchemaMapping[tableName].IsReplicaIdentityFull
```
Comment (iskakaushik) [inline flow/connectors/postgres/cdc.go:243]: looks like `isFullReplica` is unintentionally removed?

File: flow/connectors/postgres/client.go [hunk 6]
```diff
@@ -536,29 +557,28 @@ func (c *PostgresConnector) generateMergeStatement(destinationTableIdentifier st
...
 		for _, colName := range otherCols {
 			tmpArray = append(tmpArray, fmt.Sprintf("%s=src.%s", colName, colName))
 		}
 		ssep := strings.Join(tmpArray, ",")
-		quotedCols := strings.Join(unchangedColsArray, ",")
 		updateStmt := fmt.Sprintf(`WHEN MATCHED AND
-		src._peerdb_record_type=1 AND _peerdb_unchanged_toast_columns='%s'
-		THEN UPDATE SET %s `, quotedCols, ssep)
+			src._peerdb_record_type=1 AND _peerdb_unchanged_toast_columns='%s'
+			THEN UPDATE SET %s `, cols, ssep)
```
Comment (iskakaushik) [inline flow/connectors/postgres/client.go:581]: confirm that these are quoted

File: flow/connectors/postgres/cdc.go (hunk 5 above)
Comment (heavycrystal) [reply flow/connectors/postgres/cdc.go:243]: reverted

**Review by iskakaushik — APPROVED** [review]

**Pull request merged**